29.md

```diff
@@ -6,153 +6,270 @@ Relay-based Groups
-    ["r", "wss://relay.other.com/"],
-    ["h", "Marlon Brando", "wss://relay.other.com/"],
-    ["h", "James Dean", "wss://relay.other.com/"]
+    ["group", "wss://relay.example.com/"],
+    ["group", "wss://relay.other.com/'a4041f55"],
+    ["group", "wss://relay.other.com/'5d75430e"]
   ],
 }
 ```
+
+> Note: it may be worth changing this to single-letter tags so that group web of trust can be easily calculated, even if there's no member list. It also might be desirable to split each tag into primary/secondary value, so that `h` tags can be indexed as well, independent of the relay url. Something like this: `[["r", "<relay-url>"], ["h", "<group-id>", "<relay-url>"]]`
+
 ## Member lists
 
-A `kind 3xxxx` event is a member list. Anyone MAY publish one, but clients SHOULD only trust lists published by the relay's own pubkey, or by moderators mentioned in the relay's member list event.
+A `kind 39001` event is a moderator list. To be considered valid, a moderator list MUST be published by the relay's own pubkey.
 
-Lists published by regular members MAY be used for convenience in building member list indexes, but should not be trusted or relied upon.
+A `kind 39002` event is a member list. Anyone MAY publish one, but clients SHOULD only trust lists published by the relay's own pubkey, or by moderators mentioned in the relay's member list event. Member lists published by regular members MAY be used for convenience in building member list indexes, but should not be trusted or relied upon.
 
-Member lists should have the following tags:
+Member and moderator lists should have the following tags:
 
--  A `d` tag is required so that member lists can be split into pieces, but is semantically meaningless.
--  An `r` tag indicates the relay's url.
--  A `p` tag indicates a member and whether they are a moderator.
+-  A `d` tag matching the group's id.
+-  A `p` tag indicating the person's pubkey and a relay hint.
 
 ```json
 {
-  "kind": "3xxxx",
+  "kind": "39002",
   "tags": [
-    ["d", "9ce3f91a937"],
-    ["r", "wss://relay.example.com/"],
+    ["d", "<group-id>"],
+    ["p", "<pubkey>", "<relay-url>"]
     ["p", "<pubkey>", "<relay-url>"]
-    ["p", "<pubkey>", "<relay-url>", "moderator"]
   ],
 }
 ```
 
-These events SHOULD be published only to the group relay, and use a `-` tag if supported.
-
-## Content retention
-
-Relays are not obligated to retain any content. Relays MAY choose not to support any part of this NIP, for example refusing host member lists in order to protect member privacy.
+Relays MAY choose to not to publish this information, restrict what pubkeys can fetch it, or reject any member lists published by users to the relay.
 
-# Rooms
+# Migration
 
-Rooms are defined by an arbitrary string containing any text. Names SHOULD be human-readable, and less than 30 characters.
+If a group needs to be moved from one relay to another, this can be done by publishing a `kind 9009`, with the following tags:
```

Comment:
what's the goal of this? to indicate to clients that the old relay is going offline?

Reply:
Yeah, this is an escape hatch for:

- Rogue admins
- Admins who disappear
- Deplatforming
- Group forks